warp10/src/main/java/io/warp10/continuum/gts/GTSHelper.java

```diff
@@ -2190,117 +2195,459 @@ public static final GeoTimeSerie bucketize(GeoTimeSerie gts, long bucketspan, in
-        // names: array of GTS names
-        // labels: array of GTS labels
-        // ticks: array of ticks being aggregated
-        // locations: array of locations being aggregated
-        // elevations: array of elevations being aggregated
-        // values: array of values being aggregated
-        // bucket span: width (in microseconds) of bucket
-        //
+        // second case: the aggregator is cabable to process an array of List instead of an array of array
+        // it uses a special class for lists that saves a memory allocation (but lists are readonly)
+        if (aggregator instanceof WarpScriptAggregatorOnListsFunction) {
+
+          // build a structure ready to use:
+          // - read only List for ticks, values  (view of the original primitive array or BitSet)
+          // - decode elevations
+          // - decode locations
+          // - expose lists of NaN when needed
+          // [tick_of_computation,[gts_classes],[label_maps],[ticks],[latitudes],[longitudes],[elevations],[values]]
+          Object[] parms = new Object[8];
+          // name and labels can be defined here
+          parms[1] = new ArrayList<String>();
+          ((ArrayList<String>) parms[1]).add(bucketized.getName());
+          parms[2] = new ArrayList<Map>();
+          ((ArrayList<Map>) parms[2]).add(labels);
+
+          // iterate on input to find buckets
+          long currentBucketEnd;
+          int currentBucketEndPosition;
+          int currentBucketStartPosition;
+          int count;
+          Object[] aggregated = null;
+          while (i >= 0 && gts.ticks[i] > (lastbucket - bucketspan * bucketcount)) {
+            currentBucketEndPosition = i;
+            currentBucketEnd = gts.ticks[i] + (lastbucket - gts.ticks[i]) % bucketspan;
+            while (i >= 0 && gts.ticks[i] > (currentBucketEnd - bucketspan)) {
+              i--;
+            }
+            currentBucketStartPosition = i + 1;
 
-        Object[] parms = new Object[8];
+            count = currentBucketEndPosition - currentBucketStartPosition + 1;
 
-        int idx = 0;
-        parms[idx++] = bucketend;
-        parms[idx] = new String[1];
-        ((String[]) parms[idx++])[0] = bucketized.getName();
-        parms[idx] = new Map[1];
-        ((Map[]) parms[idx++])[0] = labels;
-        parms[idx++] = Arrays.copyOf(subgts.ticks, subgts.values);
-        if (null != subgts.locations) {
-          parms[idx++] = Arrays.copyOf(subgts.locations, subgts.values);
-        } else {
-          parms[idx++] = new long[subgts.values];
-          Arrays.fill((long[]) parms[idx - 1], GeoTimeSerie.NO_LOCATION);
-        }
-        if (null != subgts.elevations) {
-          parms[idx++] = Arrays.copyOf(subgts.elevations, subgts.values);
+            // tick
+            parms[0] = currentBucketEnd;
+
+            // ticks list
+            parms[3] = new COWList(gts.ticks, currentBucketStartPosition, count);
```

Comment:
refactor to reuse the same COWList instance between each iteration
same comment for other COWList and ROconstantList

Reply:
As discussed, will be too hard to explain to new WarpScript learners that can run a MACROMAPPER and let input on the stack to understand.